### DISCUSSION TITLE
Mocking services in spawned machines

### DISCUSSION THREAD

**Discussion by bhardy** [statelyai/xstate#1677] [Q&A] [answered]
I apologize if I've missed this in the docs but I've searched high and low.

I'm curious if there's a way to pass services from a parent machine to the machines that it spawns. The reason I want to do this is so that I can mock the services throughout the entire machine from 1 config.

I think the [covid-tracker](https://codesandbox.io/s/covid-state-machine-lromu?from-embed=&file=/src/App.js:378-390) is a good example, so I'll use it instead of making a new one.

If you wanted to create a test (or in my case, create a story) for App.js what would be your approach to overriding the services so they don't hit the real APIs? I'm aware that you can pass a `services` object to the `useMachine` hook (or add a `withConfig`) but I can't figure out how to get that machine to pass said services along to its spawned machines.

I have hacked out a solution by conditionally adding `devMode` to the context of the parent machine and passing that value to the spawned machines. I'm then importing both `services.js` and `mockServices.js` in both machines and conditionally choosing which function to use based on the `devMode` flag. I don't like this. I've also considered passing the desired services object directly into the parent machine's context and passing the context.services to the spawned machines, which seems cleaner (and easier to mock) but also seems like it's overly complicated.

I'm beginning to wonder if I'm either missing something obvious or if this is intentionally difficult because it's a bad pattern.

Thank you

**Update:** I've switched the parent machine so that it has a create function and I'm passing the services in there instead. This is a lot cleaner but I still wonder if there's a better solution.

**Comment by 3lrv-ryan** [comment, marked as answer]
How are your spawned machines defined? Are they already defined as `Machine`? If not, I would think you could define them as `MachineConfig` and then use `createMachine` to create a new instance with the options you want. I'm not sure that's the best approach. You could also use a factory approach where your spawned machine accepting the `services` object or the specific service instances it requires.

**Reply by bhardy** [reply]
the latter is basically what I have done now and it is working fine.

**Comment by Andarist** [comment]
Why dont use a testing framework that supports mocking modules? Just mock the very leaf modules that actually hit the network or smth like that.

**Reply by bhardy** [reply]
I think something like that could work. It seemed easier to mock at the root because my services return data and I need the mocked services to all seem like they work—if that makes any sense.
